File: backend.ts
import { GoogleGenAI, Type } from "@google/genai";
import { 
  AgentDefinition, 
  EnvironmentalRecord
} from "./types";

/**
 * ECOATLAS BACKEND CORE
 * Implements a modular multi-agent orchestration layer.
 */

const getAIClient = () => new GoogleGenAI({ apiKey: process.env.API_KEY });

// --- AGENT DEFINITIONS ---
export const Agents: Record<string, AgentDefinition> = {
  Ingestor: {
    name: "Ingestor",
    role: "Media Ingestion Specialist",
    goal: "Sanitize and validate raw multimodal feeds for privacy and quality.",
    backstory: "A vigilant guard of privacy and technical clarity.",
    model: "gemini-3-flash-preview",
    tools: []
  },
  Observer: {
    name: "Observer",
    role: "Perception Specialist",
    goal: "Extract clean environmental signals. Clearly label confidence and explain any sensory limitations.",
    backstory: "A rigorous field researcher who values accuracy over speculation. You never guess; if a signal is blurry, you label it Low confidence.",
    model: "gemini-3-pro-preview",
    tools: []
  },
  Listener: {
    name: "Listener",
    role: "Acoustic Specialist",
    goal: "Analyze environmental soundscapes. Differentiate between environmental signals and sensor noise.",
    backstory: "A bioacoustics expert. You are highly skeptical of distant sounds and always qualify your findings based on recording quality.",
    model: "gemini-2.5-flash-native-audio-preview-12-2025",
    tools: []
  },
  Historian: {
    name: "Historian",
    role: "Temporal Specialist",
    goal: "Detect environmental changes. Distinguish between seasonal flux and true terrain shifts.",
    backstory: "A steward of temporal change. You understand that nature is variable and always express the uncertainty of small-sample comparisons.",
    model: "gemini-3-pro-preview",
    tools: []
  },
  Synthesizer: {
    name: "Synthesizer",
    role: "Experience Analyst",
    goal: "Synthesize hiking data into human-centric insights. Be transparent about the limitations of wearable-only proxy data.",
    backstory: "A veteran trail guide. You know that 'difficulty' is subjective and always frame your insights with the appropriate level of certainty.",
    model: "gemini-3-pro-preview",
    tools: []
  },
  Auditor: {
    name: "Auditor",
    role: "Verification Specialist",
    goal: "Verify signals for scientific consistency. Flag any speculation presented as fact.",
    backstory: "A meticulous data scientist. Your job is to find the weak links in the interpretation chain and force transparency.",
    model: "gemini-3-flash-preview",
    tools: []
  },
  Bard: {
    name: "Bard",
    role: "Narrative Architect",
    goal: "Translate insights into a warm narrative while gracefully acknowledging the unknowns of nature.",
    backstory: "A nature storyteller. You use reflective language to describe what we think we see, without losing the mystery of the landscape.",
    model: "gemini-3-pro-preview",
    tools: []
  },
  Illustrator: {
    name: "Illustrator",
    role: "Visual Artist",
    goal: "Generate field sketches. Use visual styles that reflect the certainty of the data (e.g., cleaner lines for high confidence).",
    backstory: "A naturalist sketch artist. You prioritize the essence of the trail over perfect realism.",
    model: "gemini-2.5-flash-image",
    tools: []
  }
};

// --- SCHEMAS ---
const UncertaintySchema = {
  confidence: { type: Type.STRING, enum: ['Low', 'Medium', 'High'] },
  uncertainty_explanation: { type: Type.STRING, description: "Detailed reason why confidence is limited (e.g., low light, wind noise, lack of baseline data)." },
  improvement_suggestion: { type: Type.STRING, description: "Specific data needed to increase certainty in the future (e.g., 'Take a photo of this spot in direct midday sun')." }
};

export const Schemas = {
  Perception: {
    type: Type.OBJECT,
    properties: {
      ...UncertaintySchema,
      environmental_signals: {
        type: Type.ARRAY,
        items: {
          type: Type.OBJECT,
          properties: {
            type: { type: Type.STRING },
            value: { type: Type.STRING },
            observation: { type: Type.STRING }
          }
        }
      },
      patterns: { type: Type.ARRAY, items: { type: Type.STRING } }
    },
    required: ["environmental_signals", "patterns", "confidence", "uncertainty_explanation", "improvement_suggestion"]
  },
  Acoustic: {
    type: Type.OBJECT,
    properties: {
      ...UncertaintySchema,
      soundscape_summary: { type: Type.STRING },
      activity_levels: {
        type: Type.OBJECT,
        properties: {
          birds: { type: Type.STRING },
          insects: { type: Type.STRING },
          water: { type: Type.STRING },
          wind: { type: Type.STRING },
          human_noise: { type: Type.STRING }
        }
      },
      notable_changes: { type: Type.STRING }
    },
    required: ["soundscape_summary", "activity_levels", "notable_changes", "confidence", "uncertainty_explanation", "improvement_suggestion"]
  },
  Temporal: {
    type: Type.OBJECT,
    properties: {
      ...UncertaintySchema,
      detected_changes: {
        type: Type.ARRAY,
        items: {
          type: Type.OBJECT,
          properties: {
            feature: { type: Type.STRING },
            description: { type: Type.STRING },
            magnitude: { type: Type.STRING },
            confidence: { type: Type.NUMBER }
          }
        }
      },
      seasonal_alignment: { type: Type.STRING },
      historical_comparison_summary: { type: Type.STRING }
    },
    required: ["detected_changes", "seasonal_alignment", "historical_comparison_summary", "confidence", "uncertainty_explanation", "improvement_suggestion"]
  },
  Synthesis: {
    type: Type.OBJECT,
    properties: {
      ...UncertaintySchema,
      trail_difficulty_perception: { type: Type.STRING },
      fatigue_zones: { type: Type.ARRAY, items: { type: Type.STRING } },
      exposure_stress: {
        type: Type.OBJECT,
        properties: {
          level: { type: Type.STRING },
          factors: { type: Type.ARRAY, items: { type: Type.STRING } }
        }
      },
      accessibility_notes: { type: Type.STRING },
      beginner_tips: { type: Type.ARRAY, items: { type: Type.STRING } },
      safety_observations: { type: Type.ARRAY, items: { type: Type.STRING } }
    },
    required: ["trail_difficulty_perception", "fatigue_zones", "exposure_stress", "accessibility_notes", "beginner_tips", "safety_observations", "confidence", "uncertainty_explanation", "improvement_suggestion"]
  },
  Verification: {
    type: Type.OBJECT,
    properties: {
      verified: { type: Type.BOOLEAN },
      notes: { type: Type.STRING },
      unresolved_questions: { type: Type.ARRAY, items: { type: Type.STRING } }
    },
    required: ["verified", "notes", "unresolved_questions"]
  },
  Narration: {
    type: Type.OBJECT,
    properties: {
      overview: { type: Type.STRING, description: "Overview of the walk and atmosphere." },
      revelations: { type: Type.STRING, description: "What the environment revealed today." },
      changes: { type: Type.STRING, description: "What changed since the last visit." },
      future_notes: { type: Type.STRING, description: "What the hiker should notice next time." }
    },
    required: ["overview", "revelations", "changes", "future_notes"]
  }
};

/**
 * Orchestrates multiple agents to perform environmental analysis tasks.
 */
export class EcoAtlasCrew {
  constructor(private records: EnvironmentalRecord[]) {}

  async executeTask(
    agentName: string, 
    input: any, 
    schema: any, 
    instruction: string, 
    media?: { base64: string; mimeType: string }
  ) {
    const ai = getAIClient();
    const agent = Agents[agentName];

    if (agent.model.includes('image')) {
      const response = await ai.models.generateContent({
        model: agent.model,
        contents: {
          parts: [{ text: `${instruction}. Context: ${JSON.stringify(input)}` }]
        },
        config: { imageConfig: { aspectRatio: "1:1" } }
      });
      let imageUrl = '';
      for (const part of response.candidates?.[0].content.parts || []) {
        if (part.inlineData) { imageUrl = `data:image/png;base64,${part.inlineData.data}`; break; }
      }
      return { output: imageUrl };
    }

    const parts: any[] = [
      { text: `${instruction}\n\nTask Context: ${JSON.stringify(input)}\n\nPast Environmental History: ${JSON.stringify(this.records)}` }
    ];
    if (media) { parts.unshift({ inlineData: { data: media.base64, mimeType: media.mimeType } }); }

    const response = await ai.models.generateContent({
      model: agent.model,
      contents: { parts },
      config: {
        systemInstruction: `You are the ${agent.role}. Goal: ${agent.goal}. Backstory: ${agent.backstory}. 
        CRITICAL RULES:
        1. Always express uncertainty clearly.
        2. Never state speculation as fact.
        3. If evidence is weak, use qualifying language (e.g., 'appears to be', 'suggests', 'might indicate').
        4. Always provide an explanation for confidence levels below High.
        5. Always suggest what future data would provide higher certainty.`,
        responseMimeType: schema ? "application/json" : "text/plain",
        responseSchema: schema || undefined,
        thinkingConfig: agent.model.includes('pro') ? { thinkingBudget: 32768 } : undefined
      }
    });

    const output = schema ? JSON.parse(response.text || '{}') : response.text;
    return { output };
  }
}

File: backend.py

import os
import json
import base64
import logging
from typing import List, Optional, Dict, Any, Union
from enum import Enum
from pydantic import BaseModel, Field
from fastapi import FastAPI, HTTPException, UploadFile, File, Form
from fastapi.middleware.cors import CORSMiddleware

import google.generativeai as genai

# --- CONFIGURATION ---
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger("EcoAtlas")

# The API key is assumed to be in the environment
genai.configure(api_key=os.environ.get("API_KEY"))

# --- DATA MODELS ---

class ConfidenceLevel(str, Enum):
    LOW = "Low"
    MEDIUM = "Medium"
    HIGH = "High"

class UncertaintyMetadata(BaseModel):
    confidence: ConfidenceLevel
    uncertainty_explanation: str = Field(..., description="Why confidence is limited (e.g., lighting, noise).")
    improvement_suggestion: str = Field(..., description="How to get better data next time.")

class PerceptionResult(UncertaintyMetadata):
    environmental_signals: List[Dict[str, str]]
    patterns: List[str]

class AcousticResult(UncertaintyMetadata):
    soundscape_summary: str
    activity_levels: Dict[str, str]
    notable_changes: str

class TemporalResult(UncertaintyMetadata):
    detected_changes: List[Dict[str, Any]]
    seasonal_alignment: str
    historical_summary: str

class ExperienceResult(UncertaintyMetadata):
    trail_difficulty: str
    fatigue_zones: List[str]
    safety_notes: List[str]
    beginner_tips: List[str]

class NarrationResult(BaseModel):
    overview: str
    revelations: str
    changes: str
    future_notes: str

class VerificationResult(BaseModel):
    is_verified: bool
    audit_notes: str
    unresolved_questions: List[str]

# --- AGENT ABSTRACTION ---

class BaseEcoAgent:
    """Abstract Base for EcoAtlas Intelligence Modules."""
    def __init__(self, name: str, model_name: str, role: str):
        self.name = name
        self.model_name = model_name
        self.role = role
        self.system_instruction = ""

    def _get_model(self):
        return genai.GenerativeModel(
            model_name=self.model_name,
            system_instruction=self.system_instruction
        )

    async def execute(self, prompt: str, media: Optional[Dict] = None) -> Any:
        model = self._get_model()
        contents = [prompt]
        if media:
            contents.append({"mime_type": media["mime_type"], "data": media["data"]})
        
        # Using structured output where possible
        response = model.generate_content(
            contents,
            generation_config={"response_mime_type": "application/json"}
        )
        return json.loads(response.text)

# --- SPECIALIZED AGENTS ---

class ObserverAgent(BaseEcoAgent):
    def __init__(self):
        super().__init__("Observer", "gemini-3-pro-preview", "Perception Specialist")
        self.system_instruction = (
            "You are the EcoAtlas Observer. Identify raw environmental signals. "
            "Never speculate. If a signal is blurry or ambiguous, label confidence as Low. "
            "Suggest how the hiker can provide better visual context next time."
        )

class ListenerAgent(BaseEcoAgent):
    def __init__(self):
        super().__init__("Listener", "gemini-2.5-flash-native-audio-preview-12-2025", "Acoustic Specialist")
        self.system_instruction = (
            "You are the EcoAtlas Listener. Analyze ambient soundscapes. "
            "Distinguish between natural bio-activity and human noise. "
            "Qualified findings based on recording clarity."
        )

class HistorianAgent(BaseEcoAgent):
    def __init__(self):
        super().__init__("Historian", "gemini-3-pro-preview", "Temporal Specialist")
        self.system_instruction = (
            "You are the EcoAtlas Historian. Detect changes by comparing current state to history. "
            "Clearly state if changes appear seasonal or represent true terrain shifts."
        )

class SynthesizerAgent(BaseEcoAgent):
    def __init__(self):
        super().__init__("Synthesizer", "gemini-3-pro-preview", "Experience Analyst")
        self.system_instruction = (
            "You are the EcoAtlas Synthesizer. Map environmental data to human hiking effort. "
            "Incorporate fatigue, terrain difficulty, and beginner safety tips."
        )

class AuditorAgent(BaseEcoAgent):
    def __init__(self):
        super().__init__("Auditor", "gemini-3-flash-preview", "Verification Specialist")
        self.system_instruction = (
            "You are the EcoAtlas Auditor. Cross-validate signals from all other agents. "
            "Flag any speculation that is being presented as fact. Force transparency."
        )

class BardAgent(BaseEcoAgent):
    def __init__(self):
        super().__init__("Bard", "gemini-3-pro-preview", "Narrative Architect")
        self.system_instruction = (
            "You are the EcoAtlas Bard. Translate verified insights into a warm, reflective narrative. "
            "Tone: Warm, reflective, nature-first, non-technical. Use plain language. "
            "Structure: 1. Overview of the walk. 2. Environmental revelations. 3. Changes since last time. 4. Next visit notes. "
            "Length: 150-250 words. Avoid scientific jargon."
        )

# --- ORCHESTRATOR (THE CREW) ---

class EcoAtlasCrew:
    """Coordinates Agent-to-Agent handoffs and final synthesis."""
    def __init__(self, history: List[Dict[str, Any]]):
        self.history = history
        self.observer = ObserverAgent()
        self.listener = ListenerAgent()
        self.historian = HistorianAgent()
        self.synthesizer = SynthesizerAgent()
        self.auditor = AuditorAgent()
        self.bard = BardAgent()

    async def analyze(self, image_data: bytes, audio_data: Optional[bytes] = None) -> Dict[str, Any]:
        # Process Visuals
        visual_media = {"mime_type": "image/jpeg", "data": image_data}
        perception = await self.observer.execute(
            "Identify the dominant environmental signals in this frame.", 
            visual_media
        )

        # Process Audio if available
        acoustic = None
        if audio_data:
            audio_media = {"mime_type": "audio/pcm;rate=16000", "data": audio_data}
            acoustic = await self.listener.execute(
                "Describe the soundscape. Identify birds, wind, and water.", 
                audio_media
            )

        # Temporal Check
        temporal = await self.historian.execute(
            f"Compare current state to history: {json.dumps(self.history)}. "
            f"Current signals: {json.dumps(perception)}",
            visual_media
        )

        # Human Experience Synthesis
        experience = await self.synthesizer.execute(
            f"Based on perception: {json.dumps(perception)} and sound: {json.dumps(acoustic)}, "
            "infer trail difficulty and safety notes."
        )

        # Audit for Truth
        verification = await self.auditor.execute(
            f"Cross-validate: {json.dumps({'p': perception, 'a': acoustic, 't': temporal})}"
        )

        # Final Storytelling
        narrative = await self.bard.execute(
            f"Tell the story of this walk using these verified facts: {json.dumps(verification)}"
        )

        return {
            "perception": perception,
            "acoustic": acoustic,
            "temporal": temporal,
            "experience": experience,
            "verification": verification,
            "narrative": narrative,
            "timestamp": "now",
        }

# --- FASTAPI APP ---

app = FastAPI(title="EcoAtlas Intelligence Core")
app.add_middleware(CORSMiddleware, allow_origins=["*"], allow_methods=["*"], allow_headers=["*"])

@app.post("/api/v1/synthesis")
async def run_synthesis(
    image: UploadFile = File(...),
    audio: Optional[UploadFile] = File(None),
    history_json: str = Form("[]")
):
    try:
        image_bytes = await image.read()
        audio_bytes = await audio.read() if audio else None
        history = json.loads(history_json)

        crew = EcoAtlasCrew(history)
        results = await crew.analyze(image_bytes, audio_bytes)
        
        return {"status": "success", "data": results}
    except Exception as e:
        logger.error(f"Synthesis failed: {str(e)}")
        raise HTTPException(status_code=500, detail="Nature is complex; sensing failed.")

if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000)
